README.md

Octave / MATLAB Toolbox for GDSII Stream Format
===============================================

Ulf Griesmann, NIST, 2008 - 2016 
ulf.griesmann@nist.gov, ulfgri@gmail.com


Most functions in this toolbox are in the Public Domain (see 
Notice_and_Disclaimer.pdf), with the following exceptions:

- Boolean/clipper.hpp and Boolean/clipper.cpp are subject to the
  Boost Software license 1.0: http://www.boost.org/LICENSE_1_0.txt

- Structures/private/datamatrixmex.c is subject the GNU Public
  License version 2: http://www.gnu.org/licenses/gpl-2.0.html

- Basic/gdsio/convert_float_gcc.h is subject to the GNU Public
  License version 3: http://www.gnu.org/copyleft/gpl.html
  NOTE: This file is only used when the toolbox is compiled with
  GCC, otherwise 'convert_float_generic.h' is used instead, which
  is in the Public Domain.


New releases of the toolbox can be downloaded from:

https://sites.google.com/site/ulfgri/numerical/gdsii-toolbox


Documentation
=============
Additional documentation is available on:

https://sites.google.com/site/ulfgri/numerical/gdsii-toolbox

in a tutorial: GDSII_for_the_Rest_of_Us-<date>.pdf
The file gdsii_docs-<nn>.zip contains definitions of the GDSII file
format and example scripts for the toolbox. 


Functions
=========
Toolbox functions are grouped into the following directories:

Basic:
    Contains the low level functions for reading and writing
    of files in GDSII stream format and defines objects and
    methods for working with GDSII layouts.

Elements:
    Contains functions that return gds_element objects.

Structures:
    Contains functions that return gds_structure objects

Boolean: 
    The GDSII toolbox contains a method that performs boolean
    set operations on boundary elements. This is described in more
    detail in the file: README-Boolean / README-Boolean.pdf

Misc:
    Functions that don't return gds_* objects.
    
Scripts:
    Command line scripts for Octave that can be run directly
    from the shell prompt in a Linux / Unix environment.


Compiling
=========
This software contains several MEX functions, which must be 
compiled with a C compiler (and a C++ compiler for the Clipper
library), before the library can be used. The C compiler must be
sufficiently C99 conformant; the LCC compiler that is included with
earlier versions of MATLAB will not compile many of the mex functions
(see the MATLAB documentation for compiling external functions).

For Octave on Linux, the mex functions are compiled by executing 

$ ./makemex-octave

at the shell prompt. In MATLAB or Octave on Windows the mex functions are 
compiled by changing to the ./gds2-toolbox directory and running

>> makemex

at the MATLAB/Octave command prompt. 


Useful Stuff
============
Very good viewer and editor for GDSII files: http://www.klayout.de


Help
====
If you find a bug in the software, please send a message to 
ulf.griesmann@nist.gov or ulfgri@gmail.com and I will try to fix it.


ChangeLog

Continues on github.com

2016-November-27
- version 141
- tie up loose ends in compound gds_element indexing. 
  Compound elements can now be modified more easily 
  using Octave/Matlab index notation. 
- new poly_area method for gds_element objects to calculate
  the area of boundary elements.
- small tweak to makemex-octave script and makemex.m

2016-November-14
- version 140
- fixed the remaining bugs in the 'subsref' methods for gds_library and gds_structure 
  objects.

2016-November-07
- version 139
- much improved (faster) 'subsref' methods for gds_library and gds_structure 
  objects.
- improved the speed of 'layerinfo' methods by eliminating element indexing.
- new 'layer' method for gds_element objects to query layer/data type, which
  is faster than field indexing.

2016-November-05
- version 137/138
- return [] in 'get_element_data' if an element does not have a requested 
  property instead of throwing an error.
- improved 'subsref' methods for gds_library and gds_structure objects that
  enable more complex indexing operations. E.g. it is now possible to
  write S(10).xy to get the xy record of the 10th element in a
  structure without having to copy the element first.
- update to Clipper library version 6.4

2016-October-24
- version 136
- finally squashed bug in gdsio::now
- mxlist.h: conform to C99 (default for GCC 5)
- Update README

2016-September-08
- version 135
- improve display methods to more closely conform to other display
  methods in Octave/Matlab (e.g., indent element data).
- remove superfluous ':' in display of reference elements.    

2016-September-06
- version 134
- new methods '@gds_structure/refrename.m' and '@gds_library/srename.m'
  to structures from a library together will all references to the structure.
- fix typo in 'Contents.m'
- improve methods that return gds_element properties with error checking.
- add '@gds_element/prop' method that returns user defined element properties.

2016-September-04
- version 133
- fix reading of element properties. Was completely broken.
- make writing of element properties safer by checking if
  'attr' and 'name' fields are present in a property structure during writing.
- fixed a bug in 'gdsio.c::now'.
- 'layerinfo' method for gds_structure objects.
- change display method for gds_element objects to display the number of
  elements in a compound element in brackets after the element type.
  Same style as in structure display method.          

2016-August-28
- version 132
- make 'layerinfo.m' a method of the gds_library class.
- fix spelling error in error message in '@gds_structure/add_element.m'

2016-August-24
- version 131
- fixed a bug in '@gds_structure/add_ref.m' that prevented passing
  of cell arrays with structure names.

2016-August-17
- version 130
- fix for longstanding bug in low level list handling (Florian Demski).
- in gds_open.c: ftell returns long int, not intmax_t

2016-May-11
- version 129
- fix bug that prevented creation of paths with width < 1 user unit

2016-January-01
- version 128
- fixed a regression in Basic/funcs/layerinfo.m
- further debugging of 'bbox' methods. Seem to be usable now.

2015-December-27
- version 127
- fixed numerous bugs in bbox methods. There are  bugs
  left for layouts that use array references (aref elements)
- new methods for gds_element sref and aref object for fast retrieval
  element properties
- correct error in help text of the gds_element constructor. The
  reflection in an strans record is applied after the rotation.
- print file name when a library is written with write_gds_library.
- Basic/funcs/layerinfo.m optionally returns a vector with layers that 
  contain elements
- use 'parfor' to loop over structures in 'Basic/funcs/adjmatrix.m'.
  This has currently no effect in Octave but can speed up functions
  that require the calculation of the structure hierarcy in MATLAB.
- update 'Basic/Contents.m
- update hash functions with newer version of 'gperf'
- fix help text of 'Basic/funcs/topstruct.m'
- corrected help text of Elements/gdsii_*text.m functions.
- update the format of all m-files: 'return' at the end of 
  functions is replaced by 'end', removed ';' at the end
  of function definitions.
- update the display methods for elements, structures, and libraries.
- more informative warnings when units are not defined prior to using
  Boolean set algebra operators on elements.
 
2015-November-08
- version 126
- new bbox methods for gds_library, gds_structure, and gds_element
  objects calculate bounding boxes (probably still buggy esp. with
  aref elements).
- fixed a bug in 'gdsii_bitmap' that made it impossible to use more
  than one bitmap in a layout.
- useful error message when the arguments to the gds_element
  constructor are not supplied in property/value pairs.
- more comprehensible warning in 'poly_bool' method for
  gds_element objects.
- new 'rotate' method for path and boundary elemements that directly
  rotates the polygons in a boundary or path element.
- made function 'poly_rotz.m' accessible from everywhere.
- make default path width a little wider in gdsii_pathtext.m
- updated formatting of several methods and functions.
- renamed @gds_structure/find.m --> @gds_structure/findel.m to make
  the funcion name less generic.
- renovate the method '@gds_library/topstruct.m'. The method now
  displays the top level structure(s) when no output argument is
  present. A cell array with structure names is returned otherwise.
- cosmetic improvements to printout from '@gds_library/treeview.m'
  method.
- new method '@gds_element/sname.m' which returns the referenced
  structure name in a reference element.
- further improved '@gds_structure/find_ref.m' method.
- fix @gds_element/poly_text.m and enable rendering as boundary or
  path elements.
- make @gds_structure/find_ref.m less elegant but faster.
- better initialization of structure arrays in 'Elements/gdsii_pathtext.m' and
  'Elements/gdsii_boundarytext.m'

2015-June-15, Ulf Griesmann

Elements/gdsii_arc.m
function [pg] = gdsii_arc(arc, layer, dtype, prop, plex, elflags)
%function [pg] = gdsii_arc(arc, layer, dtype, prop, plex, elflags)
%
% gdsii_arc :  return the polygon approximation of an arc
%              or a circle (segment) as a gds_element object
% 
% arc    : structure with properties defining the arc
%          EITHER :
%             arc.r :   radius of INSIDE arc in user units. 
%                       arc.r == 0 creates a circle segment.
%             arc.c :   center coordinate of the arc; 1x2 matrix
%                       Default is [0,0].
%             arc.a1 :  start angle of arc in DEGREES
%             arc.a2 :  end angle of arc in DEGREES
%                       both angles must be from the interval [0,360]
%          OR :
%             arc.x1 :  start point of INSIDE arc
%             arc.x2 :  a point on the arc
%             arc.x3 :  end point of the arc
%
%          COMMON :
%             arc.w :   WIDTH of arc in user units
%             arc.e :   maximum approximation error in user units
%             arc.pap : polynomial approximation method. If == 0
%                       the polygon is inscribed in the arc, if == 1,
%                       the polygon has the same area
%                       as the arc, if == 2 the length of the
%                       polygon is equal to the arc
%                       length. 
%                       Default is 2 (equal length).  
% layer  : layer number. Default is 1.
% dtype  : (Optional) Data type number between 0..255. Layer number 
%          and data type together form a layer specification. 
%          DEFAULT is 0.
% prop   : (Optional) a cell array of  property number and property name pairs
%             prop.attr  : 1 .. 127
%             prop.value : string with up to 126 characters
% plex   : (Optional) plex number used for grouping of
%          elements. Should be small enough to use only the
%          rightmost 24 bits. A negative number indicates the start
%          of a plex group. E.g. -5 for the first element in the
%          plex, 5 for all the others.
% elflags: (Optional) an array (string) with the flags 'T' (template
%          data) and/or 'E' (exterior data).
%
% pg     : gds_element (boundary) with the arc boundary as a closed
%          polygon.
%

% Initial version: Ulf Griesmann, NIST, February 2011
% 

    % check arguments
    if nargin < 6, elflags = []; end
    if nargin < 5, plex = []; end
    if nargin < 4, prop = []; end
    if nargin < 3, dtype = []; end
    if nargin < 2, layer = []; end

    if isempty(layer), layer = 1; end
    if isempty(dtype), dtype = 0; end

    if not( isfield(arc, 'r') || isfield(arc, 'x1') )
        error('gdsii_arc :  incomplete arc specification.');
    end
    if ~isfield(arc, 'w')
        error('gdsii_arc :  missing arc width field.');
    end
    if ~isfield(arc, 'e')
        error('gdsii_arc :  missing arc.e, max. approximation error field.');
    end
    if ~isfield(arc, 'pap')
        arc.pap = 2;
    end
    
    % convert angles degrees --> radians
    if isfield(arc, 'a1')
        arc.a1 = pi * arc.a1 / 180;
    end
    if isfield(arc, 'a2')
        arc.a2 = pi * arc.a2 / 180;
    end    

    % calculate center, radius if 3 points are given
    if isfield(arc, 'x1')
        [arc.c, arc.r] = circle_3point(x1, x2, x3);
        arc.a1 = atan2(x1(2),x1(1));
        arc.a2 = atan2(x3(2),x3(1));
    end

    % calculate the polygon approximation
    if arc.r < 0
        error('gdsii_arc :  radius must be positive.');
    end
    if ~isfield(arc, 'c')
        arc.c = [0,0];
    end

    % first the outside arc, lowest angle to highest
    rout = arc.w + arc.r;
    [np, phi, rscal] = calc_vertex_num(rout, arc.a2 - arc.a1, arc.e, arc.pap);
    R = repmat(rscal * rout, [np,1]);  % in polar coordinates
    T = linspace(arc.a1, arc.a2, np)';
    C = repmat(arc.c, [np,1]);
    xy = pol_cart(T,R) + C;
    
    % inside arc, highest angle to lowest
    if arc.r > 0                % arc segment
        R = repmat(rscal * arc.r, [np,1]);
        xy = [xy; pol_cart(T(end:-1:1),R) + C];
    else                        % circle segment
        if mod(arc.a2 - arc.a1, 2*pi)
            xy(end+1,:) = arc.c;
        end
    end

    % close the polygon
    if any(xy(end,:) ~= xy(1,:))
        xy(end+1,:) = xy(1,:);
    end

    % return it
    pg = gds_element('boundary','xy',xy, 'layer',layer, 'dtype',dtype, ...
                     'prop',prop, 'plex',plex, 'elflags',elflags);
end

%---------------------------------------------------------

function XY = pol_cart(T,R);
% convert from polar to cartesian coordinates
    
    XY = [R.*cos(T), R.*sin(T)];

end

%---------------------------------------------------------
   
function [np, phi, Rfac] = calc_vertex_num(R, ang, E, pap);
% calculate number of vertices needed to approximate an arc by 
% a polygon
%
% R :    radius of the arc
% ang :  angle subtended by arc
% E :    maximum permissible approximation error
% pap :  polynom approximation method: 0 = polygon inscribed in
%        arc; 1 = arc area is equal to polygon area; 2 = polygon
%        length is equal to arch length
% np :   number of point in polygon approximation
% phi :  angle increment between vertices
% Rfac:  radius scale factor, depends on pap
%
    phi  = acos(1 - E / R);
    np   = ceil(ang / phi);  % integer number of vertices
    phi  = ang / (np - 1);   % corrected angle
    
    switch pap
      case 0  % polygon is inscribed in arc
        Rfac = 1;
        
      case 1  % polygon area equals arc area
        Rfac = sqrt( phi / sin(phi) );
        
      case 2  % polygon length equals arc length
        Rfac = 0.5 * phi / sin(phi/2);
        
      otherwise
        error('unknown circle approximation method.');
    end

end

%---------------------------------------------------------

function [C,R] = circle_3point(x1, x2, x3);
%function [C,R] = circle_3point(x1, x2, x3);
%
% circle_3point :  calculate center and radius of a circle
%                  defined by three points on the circle
%
% x1,x2,x3 :  three points (coordinate pairs)
% C :         Center of circle
% R :         Radius of circle
%

% initial version, Ulf Griesmann, NIST, September 2011

    if nargin < 3
        error('must have three arguments.');
    end
    
    A = [x3(1) - x1(1), x3(2) - x1(2); ...
         x3(1) - x2(1), x3(2) - x2(2) ];
    b = [0.5*(x3(1)^2 - x1(1)^2 + x3(2)^2 - x1(2)^2 ); ...
         0.5*(x3(1)^2 - x2(1)^2 + x3(2)^2 - x2(2)^2 )];
    C = A\b;
    R = sqrt(sum((x1-C).^2));
    
end
